Bounty: Mazda

Jafaral opened the pull request:
Migrating PR #762 to master

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

geohot commented:
Does this still have the 10 second steering lockout?

bugsy924 commented:
Yes, alleviated with a weight on the steering

On Jan 24, 2020 6:44 AM, George Hotz <notifications@github.com> wrote:

Does this still have the 10 second steering lockout?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/commaai/openpilot/pull/988?email_source=notifications&email_token=AHMDMSFGBVDQ4X2EJGGQU7TQ7HXTJA5CNFSM4KJYMPY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJYTA4Q#issuecomment-577843314>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHMDMSA7KPMYHS3AXAQRMRDQ7HXTJANCNFSM4KJYMPYQ>.

Jafaral commented:
@geohot yes, and it is 5 seconds actually.  It is different from the ford lockout as I explained in #762. As @Bugsy#7983 the problem is mitigated with a weight on the steering wheel.  In my case I use a steering wheel cover with some weight hidden along the curve.  Didn't have issues since.

geohot commented:
So this doesn't work for something we can ship. If we want to merge this, we'd have to detect and alert the user whenever they are locked out. Can you add an alert for this?

Though I hear good news for 2019 Mazda 3, that there's no lockout?

Jafaral commented:
Fair, I can easily add an alert and maybe issue a soft disable, would that work ?

correct, Mazda3 doesn't seem to have the issue, but it still needs some work porting. I have been focusing on getting the current Gen Mazda cars officially supported before I put more effort into Mazda3.

I will the lockout alert and update the PR.

Jafaral commented:
I have worked on adding alerts for steer lockout and low speed and working on improving various aspects of the port. Will be updating this PR in the new few days.

Jafaral commented:
@geohot I have pushed a few updates to this PR with a few improvements and also to address several situations including  your comment above. Notable new changes:
     
      * Do not engage if LKAS is not available at low speeds. This happens anytime  the speed dips below 28mph and doesn't get enabled until the speed hits 32mpg. If the user tried to engage they get "OpenPilot Unavailable, Speed Two Low" 

      * If OP is engaged and the speeds dips below 28mpg where LKAs becomes unavailable the user see a continuous "TAKE CONTROL: Steer Unavailable Below 28MPH"  warning. OP doesn't disengage in this case but the warning will be shown as long as the LKAs is not available. Once the speed requirement is met the warning goes away and OP resumes normal operation. 

    * If the user have their hands off the steering wheel and the steer lockout timeout is triggered,  a Warning is issued "TAKE CONTROL: Steering is Temporarily Unavailable". We can create a specific warning for cars with lockout timeout if we want to make it more obvious way it is not available. The stock car system itself also issues a HUD warning and an audible alert in that situation, so the user is very aware of what  is going on.

   * Disengage on gas press which is required by Comma AI safety guidelines. break/seatbelt/door disengagement were already there. 

   * Refactored a lot of the car params/constants and moved them to values.py

   * Improved the handling of cruise buttons presses to give a better user feedback
   
   * Added CX-2019 GT fingerprint as a couple of new Mazda users have that car and use it with Comma 2

   The only remaining issue is that sometimes I get steer limit warnings even when driving on straight  lines. More tuning is needed for that.

Please let me know if you have any comments or questions!

Jafaral commented:
I did write a guide for Mazda user covering all aspects of this port.  I will update the guide with screen shots of the various warnings as examples when I get a chance:

https://medium.com/@to.jafar/how-to-setup-openpilot-on-mazda-3eb54c62fdc5

Jafaral commented:
I update the Mazda guide above with example warnings. Will add one for hands off warning when I get one.

geohot commented:
The expectation from every car in openpilot is that it be able perform on the highway without many disengagements ( referenced in #1060 ), and unfortunately with a 10 second lockout that's unlikely. I want to be clear on this policy going forward, sorry for being indecisive.

If some cars don't have the lockout, we can upstream them. I still think the warning is a good idea, and we'll keep this PR open in hopes future Mazda's don't have the lockout, or a way is found to bypass it.

Jafaral commented:
@geohot, I understand.  I'd like to point out that on the highway I hardly get any disengagement for long stretches, 10+ minutes, with a neatly hidden weight under a steering wheel cover. All Mazda users are using this trick.

geohot commented:
Let's merge this as a dashcam only port for now.

Jafaral commented:
@geohot Thanks for the update!

@pd0wm aside from updating the PR to address the merge conflicts, please let me know what other things you want added/dropped to make it work in dashcam mode only.

geohot commented:
Can you fix test car models?

pd0wm commented:
Just set `    ret.dashcamOnly = True` in interface.py. (https://github.com/commaai/openpilot/blob/master/selfdrive/car/ford/interface.py#L22). Also make sure the route gets marked as such in test_car_models. Just copy what ford does.

Also there have been some refactors lately in carstate/carinterface. Can you make sure you follow the same conventionts:
 - Build carState struct in carstate.py, not in interface. That save a lot of copies
 - Use the `create_common_events` method to avoid some duplication in the regular events like door, seatbelt, etc.
 - Use the kalman filter from CarStateBase
 - I see the gear is not parsed. Make sure it disengages when going in reverse.

Jafaral commented:
@geohot  @pd0wm will do!  I will update the PR and test in the next few days.

Jafaral commented:
Finally got some time to move a bit on this, still more work to do. Will update when I get more time to work on it.

Jafaral commented:
@geohot
  test is working now

 @pd0wm 

  * Code was refactored following the new convention.
  * Gear is parsed
  * gas/brake events including pressure are all added 
  * Fingerprints were added to cover many of the Mazda cars on the road now
  * refactored the code to prepare for future updates for the new gen Mazda  such as Mazda 3

Remaining items:
    * Force dashmode: I'm keeping this as the last step after everything else is working/approved
    * Disengage on gas-press , this was working before, missing something in my refactoring
    * Fix process_replay test, the drive is synced by you but I still get this error:
```AssertionError: Failed to download /tmp/openpilot/selfdrive/test/process_replay/32a319f057902bb3|2020-04-27--15-18-58--2_controlsd_b33259e2dae9852c58b0507dbb98fb01d8b884b8.bz2```

If you have any comments please let me know.

quillford commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,214 @@
+
+    self.speed_kph =  ret.vEgoRaw  // CV.KPH_TO_MS
+
+    gear = int(cp.vl["GEAR"]['GEAR']) >> 1
+
+    if gear >= 1 & gear <= 5:
+      ret.gearShifter = GearShifter.drive
+    elif gear == 0:
+      ret.gearShifter = GearShifter.park
+    elif gear == 6:
+      ret.gearShifter = GearShifter.reverse
+    elif gear == 7:
+      ret.gearShifter = GearShifter.neutral
+    else:
+      ret.gearShifter = GearShifter.unknown
```
Why the shift for the gear value? Also seems like a pretty big range for drive. If you define these values in the DBC, you can use [parse_gear_shifter](https://github.com/commaai/openpilot/blob/master/selfdrive/car/interfaces.py#L158) like the other cars.

Jafaral replied:
Shift because I found that the low bit is not used, i.e shouldn't be part of the signal. I already opened a PR to fix that (https://github.com/commaai/opendbc/pull/254) once that is merged the shift can be dropped. 1 through 5 are gear positions for speeds  1 to 5.  I did see this that function, do I fill in letters instead and leave the mapping to the function?

quillford replied:
You have to define the gear values in the DBC like this https://github.com/commaai/opendbc/blob/master/toyota_prius_2017_pt_generated.dbc#L401

Jafaral replied:
Do I edit the file manually or can this be done via cabana?

quillford replied:
I don't think cabana can do this. I'd just do it manually

Jafaral replied:
here is what I came up with: 
```
VAL_ 552 GEAR 0 "P" 1 "D" 2 "D" 3 "D" 4 "D" 5 "D" 6 "R" 7 "N";
```
There might be a better signal to  track the "stick" state, but I haven't found that one if it does exist.

Jafaral replied:
Does this work?

quillford replied:
Yeah, that should work. What do you mean by stick state? If you mean the actual gear position, I'd just put it in a comment

quillford commented on selfdrive/car/mazda/interface.py:
```diff
@@ -0,0 +1,114 @@
+      self.low_speed_alert = False
+
+    # events
+    events = self.create_common_events(ret)
```
this is already called in line 85

Jafaral replied:
good catch!

quillford commented on selfdrive/car/mazda/values.py:
```diff
@@ -0,0 +1,69 @@
+
+}
+
+class ECU:
+  CAM = 0
+
+ECU_FINGERPRINT = {
+  ECU.CAM: [579],   # steer torque cmd
+}
```
don't need to make a new class. you can use car.CarParams.Ecu from cereal like this https://github.com/commaai/openpilot/blob/master/selfdrive/car/toyota/values.py#L64

Jafaral replied:
fixed!

quillford commented on selfdrive/car/mazda/interface.py:
```diff
@@ -0,0 +1,114 @@
+    if self.low_speed_alert:
+      events.append(create_event('belowSteerSpeed', [ET.WARNING]))
+
+    if self.CS.steer_lkas.handsoff:
```
If you use `self.steer_warning` in carstate.py, `create_common_events` will create the event for you.

Jafaral replied:
noted!

quillford replied:
sorry, I actually just changed this. set carState.steerWarning instead  https://github.com/commaai/openpilot/pull/1444

Jafaral replied:
double noted!

quillford commented on selfdrive/car/mazda/mazdacan.py:
```diff
@@ -0,0 +1,57 @@
+    else:
+      csum = csum + 256
+
+  csum = csum % 256
```
If the checksum isn't specific to this message, maybe give it its own function

Jafaral replied:
I doubt the checksum will work in other messages. It is all over the place and  specific to the signals of this message!

quillford commented:
@Jafaral process replay needs both the drive synced and the reference logs it compares against. That error means the reference logs are missing from the CI bucket

Jafaral commented:
@quillford The drive 32a319f057902bb3|2020-04-27--15-18-58 is used with test_car_models as well and I was getting an error there until @pd0wm synced it. After the sync, the test_Cat_models went away but the the error for the segment within that drive here persist, not sure what that means.

Jafaral commented:
@quillford Aside from parsing the gear, all other comments are addressed. Will test drive this tomorrow. Will wait for the dbc updates to refactor gear parsing.

Thanks for taking the  time to look at this!

quillford commented on selfdrive/car/mazda/carcontroller.py:
```diff
@@ -0,0 +1,35 @@
+
+class CarController():
+  def __init__(self, dbc_name, CP, VM):
+    self.steer_idx = 0
```
unused

quillford commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,219 @@
+      ret.cruiseState.speed = 0
+
+
+    self.brake_error = False
```
unused

quillford commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,217 @@
+    elif self.speed_kph < LKAS_LIMITS.DISABLE_SPEED and not self.low_speed_lockout:
+      self.low_speed_lockout = True
+
+    if (self.low_speed_lockout or self.steer_lkas.block) and self.speed_kph < LKAS_LIMITS.DISABLE_SPEED:
```
self.steer_lkas block can just be local to the function. don't see it used anywhere else

quillford commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,217 @@
+
+GearShifter = car.CarState.GearShifter
+
+class STEER_LKAS():
+  def __init__(self):
+    self.block = 1
+    self.track = 1
+    self.handsoff = 0
```
doesn't seem like this is needed. block should be local to CS.update, tack is unused, and handsoff will be replaced by steerWarning

quillford commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,217 @@
+    ret.steeringTorque = cp.vl["STEER_TORQUE"]['STEER_TORQUE_SENSOR']
+    ret.steeringPressed = abs(ret.steeringTorque) > LKAS_LIMITS.STEER_THRESHOLD
+
+    self.steer_torque_motor = cp.vl["STEER_TORQUE"]['STEER_TORQUE_MOTOR']
```
`ret.steeringTorqueEps`

quillford commented on selfdrive/car/mazda/interface.py:
```diff
@@ -0,0 +1,103 @@
+    # TODO: button presses
+    ret.buttonEvents = []
+
+    if ret.cruiseState.enabled and self.CS.lkas_speed_lock:
+      self.low_speed_alert = True
+    else:
+      self.low_speed_alert = False
```
unused

quillford commented on selfdrive/car/mazda/interface.py:
```diff
@@ -0,0 +1,103 @@
+ButtonType = car.CarState.ButtonEvent.Type
+
+class CarInterface(CarInterfaceBase):
+  def __init__(self, CP, CarController, CarState):
+    super().__init__(CP, CarController, CarState)
+
+    self.low_speed_alert = False
```
`self.low_speed_alert` is unused. this can all be deleted

quillford commented:
@Jafaral I merged your opendbc PR, so you can fix the gear parsing now.

Before this gets merged, the mazda panda safety should have tests. You can use the [toyota safety test](https://github.com/commaai/panda/blob/master/tests/safety/test_toyota.py) as an example. At a minimum, it should include all the tests from `common.PandaSafetyTest`.

Jafaral commented:
Thanks @quillford , it also needs to be synced to the OP's git submodules before it actually works, right?

I will address your latest comment and prepare for gear parsing. Will do panda tests as well.

One of my changes last night caused controlsd crash this morning. I will track that down. Can  you please find out why  process_replay is crashing? That  should help me catch runtime errors before I roll them on to the car. I can provide a new segment ID if that fixes the issue.

Thanks Again!

Jafaral commented:
Just pushed a Dashcam only commit as requested to finalize this PR.

adeebshihadeh commented on selfdrive/car/mazda/values.py:
```diff
@@ -0,0 +1,69 @@
+  STEER_DRIVER_FACTOR = 1         # from dbc
+
+class CAR:
+#  CX3 = "Mazda CX-3"
```
Seems like these are supported, but just treated as a CX5. Either break them out into separate cars, or delete these commented lines.

Jafaral replied:
They all have identical CAN buses and control, and even overlapping fingerprints. Many of these were added recently as more Mazda cars came online. I was planning on adding blocks for each car but hasn't been a priority so far

adeebshihadeh commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,194 @@
+                cp.vl["CRZ_BTNS"]['SET_P'],
+                cp.vl["CRZ_BTNS"]['SET_M']]):
+      self.acc_active = True
+      self.cruise_speed = ret.vEgoRaw
```
There should be a signal for set speed in the can data. Every other car has this.

Jafaral replied:
I'm sure there is! no long control, so it wasn't as important to dig out, will try to find it.

adeebshihadeh commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,194 @@
+      self.cruise_speed = ret.vEgoRaw
+
+    ret.cruiseState.available = cp.vl["CRZ_CTRL"]['CRZ_ACTIVE'] == 1
+    if not ret.cruiseState.available:
+      self.acc_active = False
```
doesn't need a condition

Jafaral replied:
why not ?

Jafaral replied:
Only want to sync if cruiseState.available is false. acc_active can be already false but available is true

adeebshihadeh commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,194 @@
+        self.low_speed_lockout = False
+        self.low_speed_alert = False
+
+      if handsoff:
+        ret.steerWarning = True
```
doesn't need a condition

Jafaral replied:
Why not ?

adeebshihadeh replied:
`ret.steerWarning = handsoff`

Jafaral replied:
ah yes

adeebshihadeh commented:
Can you add Mazda to the README? Thanks for all the work on this port, it's looking pretty good now.

adeebshihadeh commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,190 @@
+    self.acc_active_last = False
+    self.low_speed_lockout = True
+    self.low_speed_alert = False
+    self.lkas_on = False
```
Let's do the same rename we did in panda: `lkas_allowed`

adeebshihadeh commented on selfdrive/car/mazda/carcontroller.py:
```diff
@@ -0,0 +1,44 @@
+    self.packer = CANPacker(dbc_name)
+    self.steer_rate_limited = False
+
+  def update(self, enabled, CS, frame, actuators, pcm_cancel_cmd, visual_alert,
```
Remove unused args

Jafaral replied:
I just added those last night hoping to build on top of them as we go!

adeebshihadeh replied:
I'd prefer to only include what's needed now. Easy enough to add later.

adeebshihadeh commented on selfdrive/car/mazda/values.py:
```diff
@@ -0,0 +1,63 @@
+    # CX-5 2017 GT
+    {
+      64: 8, 70: 8, 80: 8, 117: 8, 118: 8, 120: 8, 121: 8, 130: 8, 134: 8, 145: 8, 154: 8, 155: 8, 157: 8, 158: 8, 159: 8, 253: 8, 304: 8, 305: 8, 357: 8, 358: 8, 359: 8, 512: 8, 514: 8, 515: 8, 529: 8, 533: 8, 535: 8, 539: 8, 540: 8, 541: 8, 542: 8, 543: 8, 552: 8, 576: 8, 577: 8, 578: 8, 579: 8, 580: 8, 581: 8, 582: 8, 605: 8, 606: 8, 607: 8, 608: 8, 628: 8, 832: 8, 836: 8, 863: 8, 865: 8, 866: ...
+  },
```
indent

adeebshihadeh commented on selfdrive/car/mazda/carcontroller.py:
```diff
@@ -0,0 +1,44 @@
+        # Send Resume button at 2hz if we're engaged at standstill to support full stop and go!
+        # TODO: improve the resume trigger logic by looking at actual radar data
+        can_sends.append(mazdacan.create_cancel_acc(self.packer, CS.CP.carFingerprint, Buttons.RESUME))
+    else:
```
This should be like the other cars where we cancel on `pcm_cancel_cmd`

Jafaral replied:
Sometimes we may need to send a couple of msgs to get acc cancel, would switching to pcm_cancel_cmd cause an issue there?

adeebshihadeh replied:
No, it should behave the same in that regard

adeebshihadeh commented on selfdrive/car/mazda/carcontroller.py:
```diff
@@ -0,0 +1,44 @@
+      new_steer = int(round(actuators.steer * SteerLimitParams.STEER_MAX))
+      apply_steer = apply_std_steer_torque_limits(new_steer, self.apply_steer_last,
+                                                  CS.out.steeringTorque, SteerLimitParams)
+      self.steer_rate_limited = new_steer != apply_steer
```
This isn't updated when not enabled

adeebshihadeh commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,190 @@
+        self.low_speed_lockout = False
+        self.low_speed_alert = False
+
+      ret.steerWarning = handsoff
```
handsoff isn't used anywhere else, just directly set ret.steerWarning

adeebshihadeh commented on selfdrive/car/mazda/carstate.py:
```diff
@@ -0,0 +1,190 @@
+    ret.gas = cp.vl["ENGINE_DATA"]['PEDAL_GAS']
+    ret.gasPressed = ret.gas > 0
+
+    # No steer if block signal is on
+    # block = cp.vl["STEER_RATE"]['LKAS_BLOCK'] == 1
```
Unused, remove

adeebshihadeh commented on selfdrive/car/mazda/mazdacan.py:
```diff
@@ -0,0 +1,104 @@
+
+def create_button_cmd(packer, car_fingerprint, button):
+
+  if button == Buttons.CANCEL:
+    can = 1
+    res = 0
+  elif button == Buttons.RESUME:
+    can = 0
+    res = 1
+  else:
+    can = 0
+    res = 0
```
```
can = button == Buttons.CANCEL
res = button == Buttons.RESUME
```
